### DISCUSSION TITLE
Function Overloading

### DISCUSSION THREAD

**Discussion by Deytooh** [naverz/zepeto-studio-global#19] [Q&A]
Hello ~ It's me, again 🌞 

I would like to know how to do function overloading in typescript ?

**Comment by jhslee** [comment]
Hello again @Deytooh! 👋 

Yes of course!! Happy to help 😁 

So function overloading is a bit tricky in typescript. 

In the example below: 

```
//Definitions
Print(a: string);
Print(a: number);
Print(a: string, b: string);
Print(a: number, b: string);
Print(a: number, b: number); //Error: Invalid Definition 

//Implementation
Print(a: any, b?: string) {
     if (b != undefined)
          console.log(a.toString() + " " + b.toString());
     else
          console.log(a.toString());
}


Start() {
     this.Print(1, "Test");
     this.Print("Another", "Test");
     this.Print(3);
     this.Print("Final Test");
     this.Print(Vector3.one); //Invalid Call
}
```

You'll notice 4 different function definitions, then the implementation below them. Because Javascript is originally untyped, when typescript compiles to Javascript, we need to ensure that the function declarations are compatible.  

We have our implementation parameters: `Print(a: any, b?: string)`. 

Compatible Definitions:
`Print(a: string);` matches our implementation because our `a` is of type `any` and our `b` is optional. 
`Print(a: number);` matches our implementation because our `a` is of type `any` and our `b` is optional.
`Print(a: string, b: string);` matches our implementation because our `a` is of type `any` and our `b` is of type `string`
`Print(a: number, b: string);` matches our implementation because our `a` is of type `any` and our `b` is of type `string`

Incompatible Definitions: 
`Print(a: number, b: number);` will be an invalid definition because our `b` is of type `string`

Incompatible Calls: 
`this.Print(Vector3.one);` will be an invalid call to print because even through our `a` is of type `any`, we did not define the `Print(a: Vector3);`


The resulting output (Removing Error lines):
![image](https://user-images.githubusercontent.com/105688184/172964806-83c4db28-011a-448f-a8ac-7440d6b74637.png)

**Reply by Deytooh** [reply]
Helloooo !! 🔥 

Oh ! Great !! I was not sure about how to do that in the Unity environment but great, thank you very much 🔥 It is a great help !!
